sqlglot/diff.py

```diff
@@ -108,44 +104,47 @@ def diff(
         expression tree into the target one.
     """
     matchings = matchings or []
-    matching_ids = {id(n) for pair in matchings for n in pair}
 
     def compute_node_mappings(
-        original: exp.Expression, copy: exp.Expression
+        old_nodes: tuple[exp.Expression, ...], new_nodes: tuple[exp.Expression, ...]
     ) -> t.Dict[int, exp.Expression]:
         node_mapping = {}
-        for old_node, new_node in zip(
-            reversed(tuple(original.walk())), reversed(tuple(copy.walk()))
-        ):
-            # We cache the hash of each new node here to speed up equality comparisons. If the input
-            # trees aren't copied, these hashes will be evicted before returning the edit script.
+        for old_node, new_node in zip(reversed(old_nodes), reversed(new_nodes)):
             new_node._hash = hash(new_node)
-
-            old_node_id = id(old_node)
-            if old_node_id in matching_ids:
-                node_mapping[old_node_id] = new_node
+            node_mapping[id(old_node)] = new_node
```

Comment:
I guess if the same node reference appears more than once in the original ASTs then this line would overwrite the first node in the `node_mapping` and it may silently cause problems.

This is probably true today as well, just pointing it out. It would be helpful if we warned, because I can imagine that debugging these scenarios in the future will be very gnarly. We could maybe avoid warning if the ids of the top-level nodes match, and just short-circuit by returning an empty list.